### ISSUE TITLE
Add filename to deprecated pluginlib warning

### ISSUE THREAD

**Issue by trainman419** [ros/pluginlib#14]
When starting a large project, it is difficult to determine where the pluginlib macro warnings are coming from. It would be useful if the macro included the filename where it was invoked.

**Comment by mirzashah** [comment]
Yes good idea, will make a note to add that in.

Though the thrill of the hunt for mystery macro warnings is spectacular.

**Comment by danepowell** [comment]
Mmm... the thrill of the hunt, I guess that's one way to look at it :)

Any idea how to implement this, or tips for tracking down where this warning comes from?

It might actually make more sense to implement this in the class_loader library.

**Comment by mirzashah** [comment]
@danepowell  The patch is very easy, but I kept running into some strange problem on my computer where ROS logging was no longer working, so I couldn't tell if the patch worked. I never had the chance to figure it out and put it on the back burner.

Essentially all that is required is use of the `__FILE__`macro to https://github.com/ros/pluginlib/blob/groovy-devel/include/pluginlib/class_list_macros.h#L49 and to https://github.com/ros/pluginlib/blob/groovy-devel/include/pluginlib/class_list_macros.h#L58. If either you or @trainman419 could try that out and send me a pull request, I'd greatly appreciate it.

The changes:

```
#define PLUGINLIB_REGISTER_CLASS(class_name, class_type, base_class_type) \
  CLASS_LOADER_REGISTER_CLASS_WITH_MESSAGE(class_type, base_class_type, "pluginlib WARNING: PLUGINLIB_REGISTER_CLASS is deprecated, please use PLUGINLIB_EXPORT_CLASS instead. You can run the script 'plugin_macro_update' provided with pluginlib in your package source folder to automatically and recursively update legacy macros. Base = base_class_type, Derived = derived_class_type")
```

should be changed to

```
#define PLUGINLIB_REGISTER_CLASS(class_name, class_type, base_class_type) \
  CLASS_LOADER_REGISTER_CLASS_WITH_MESSAGE(class_type, base_class_type, "pluginlib WARNING: In file __FILE__, PLUGINLIB_REGISTER_CLASS is deprecated, please use PLUGINLIB_EXPORT_CLASS instead. You can run the script 'plugin_macro_update' provided with pluginlib in your package source folder to automatically and recursively update legacy macros. Base = base_class_type, Derived = derived_class_type")
```

and also PLUGINLIB_DECLARE_CLASS becomes:

```
#define PLUGINLIB_DECLARE_CLASS(pkg, class_name, class_type, base_class_type) \
  CLASS_LOADER_REGISTER_CLASS_WITH_MESSAGE(class_type, base_class_type, "pluginlib WARNING: In file __FILE__,  PLUGINLIB_DECLARE_CLASS is deprecated, please use PLUGINLIB_EXPORT_CLASS instead. You can run the script 'plugin_macro_update' provided with pluginlib in your package source folder to automatically and recursively update legacy macros.  Base = base_class_type, Derived = derived_class_type")            
```

**Comment by danepowell** [comment]
Thanks @mirzashah . I applied the fix, but in a bit of cosmic irony, I'm unable to say if it works. Because the fix is in a header file, I need to recompile the offending package to see if it tells me what the offending package is :)

I'll keep hunting or find another way of testing and report back.

**Comment by mirzashah** [comment]
@danepowell - Unfortunately it's difficult to implement plugins in C/C++ without the use of macros due to the way the language works.

You can just write a simple program or even use the exisiting unit tests in pluginlib to test.

For the latter, to build and run the pluginlib unit tests:
catkin_make run_tests_pluginlib

Change the use of PLUGINLIB_EXPORT_CLASS in the test plugin to one ofthe deprecated versions and see if it invokes the warning.

**Comment by dirk-thomas** [comment]
Since class_loader is about to be released soon is there any chance we can land this helpful information before that?

**Comment by danepowell** [comment]
I opened a pull request: https://github.com/ros/pluginlib/pull/16

(I'm not sure if there's a way to link it to this issue)

**Comment by tfoote** [comment]
If you use hub to open the pull request you can but otherwise it's too difficult to be useful.

**Comment by dirk-thomas** [comment]
@mirzashah Can you modify the macro in _class_loader_ in the same way as _pluginlib_? Than I will take over and release both packages later today.

**Comment by mirzashah** [comment]
@dirk-thomas The problem specified only applies to pluginlib. The warning comes from using the deprecated macros PLUGINLIB_REGISTER_CLASS and PLUGINLIB_DECLARE_CLASS instead of the new PLUGINLIB_EXPORT_CLASS. class_loader has its own macro CLASS_LOADER_REGISTER_CLASS which is used in the implementation of the aforementioned pluginlib macros.

**Comment by dirk-thomas** [comment]
Oh, okay, I misread the earlier conversation with all these long macro names.

Great, I will go ahead and release both packages soonish. Thanks.

**Comment by dirk-thomas** [comment]
Released into Groovy as well as Hydro.

**Comment by trainman419** [comment]
This isn't working. I'm getting: 

```
pluginlib WARNING: In file __FILE__ PLUGINLIB_DECLARE_CLASS is deprecated, please use PLUGINLIB_EXPORT_CLASS instead. You can run the script 'plugin_macro_update' provided with pluginlib in your package source folder to automatically and recursively update legacy macros.  Base = base_class_type, Derived = derived_class_type
```

I'm fairly certain that the **FILE** macros don't work inside of a string constant.

**Comment by dirk-thomas** [comment]
@mirzashah Please take a look and provide a modified patch to make it work. Then we need to re-release into Groovy and Hydro.

**Comment by mirzashah** [comment]
Fixed in ddfc5864fa5e6d0b41f6c5a0fab4aa71e863f207
